### PR TITLE
[WIP] Windows/CMake build

### PR DESCRIPTION
**Pull request by beauby**

### CODE REVIEW

File: faiss/Clustering.cpp [hunk 1]
```diff
@@ -97,7 +98,7 @@ idx_t subsample_training_set(
 )
 {
     if (clus.verbose) {
-        printf("Sampling a subset of %ld / %ld for training\n",
+        printf("Sampling a subset of %zd / %" PRId64 " for training\n",
```
Comment (mdouze) [inline faiss/Clustering.cpp:101]: what is the difference between zd and PRId64 ?

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  target_sources(faiss PRIVATE
+    faiss/OnDiskInvertedLists.cpp)
+endif()
+
+# Handle `#include <faiss/foo.h>`.
+target_include_directories(faiss PUBLIC
+  $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>)
+
+set_property(TARGET faiss PROPERTY POSITION_INDEPENDENT_CODE ON)
+target_compile_definitions(faiss PRIVATE FINTEGER=long)
```
Comment (mdouze) [inline CMakeLists.txt:81]: it is FINTEGER=int on Linux

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  HOMEPAGE_URL "https://github.com/facebookresearch/faiss"
+  LANGUAGES CXX)
+
+set(CMAKE_CXX_STANDARD 11)
+
+option(PYTHON_VERSION "Python version to pyfaiss build against.")
+option(FAISS_ENABLE_GPU "Enable support for GPU indexes." ON)
+
+add_library(faiss
+  faiss/IndexLSH.cpp
```
Comment (mdouze) [inline CMakeLists.txt:20]: So you moved all the code to a faiss/ subdirectory?

File: faiss/IndexBinary.cpp [hunk 4]
```diff
@@ -70,7 +71,7 @@ void IndexBinary::search_and_reconstruct(idx_t n, const uint8_t *x, idx_t k,
 }
 
 void IndexBinary::display() const {
```
Comment (mdouze) [inline faiss/IndexBinary.cpp:73]: I think we removed the display function for the float indexes

File: faiss/MetaIndexes.cpp [hunk 5]
```diff
@@ -319,7 +321,7 @@ void IndexSplitVectors::search (
                     distances[j] += distances_i[j];
                 } else {
                     labels[j] = -1;
-                    distances[j] = 0.0 / 0.0;
+                    distances[j] = std::numeric_limits<double>::quiet_NaN();
```
Comment (mdouze) [inline faiss/MetaIndexes.cpp:324]: it's a float

File: faiss/impl/index_read.cpp [hunk 6]
```diff
@@ -70,7 +73,7 @@ namespace faiss {
 #define READVECTOR(vec) {                       \
         long size;                            \
         READANDCHECK (&size, 1);                \
-        FAISS_THROW_IF_NOT (size >= 0 && size < (1L << 40));  \
+        FAISS_THROW_IF_NOT (size >= 0 && size < (1LL << 40));  \
```
Comment (mdouze) [inline faiss/impl/index_read.cpp:76]: better to do explicit cast (uint64_t)1 << 40.

File: faiss/impl/lattice_Zn.cpp [hunk 7]
```diff
@@ -115,12 +131,12 @@ long repeats_encode_64 (
         uint64_t tosee = ~coded;
         for(;;) {
             // directly jump to next available slot.
-            int i = __builtin_ctzl(tosee);
-            tosee &= ~(1UL << i) ;
```
Comment (mdouze) [inline faiss/impl/lattice_Zn.cpp:119]: explicit cast

File: faiss/impl/lattice_Zn.cpp [hunk 8]
```diff
@@ -341,9 +357,9 @@ ZnSphereSearch::ZnSphereSearch(int dim, int r2): dimS(dim), r2(r2) {
 }
 
 float ZnSphereSearch::search(const float *x, float *c) const {
-    float tmp[dimS * 2];
-    int tmp_int[dimS];
-    return search(x, c, tmp, tmp_int);
+    std::vector<float> tmp(dimS * 2);
```
Comment (mdouze) [inline faiss/impl/lattice_Zn.cpp:360]: a bit frustrating to do dynamic allocs here. Another way of doing it?

File: faiss/tests/test_dealloc_invlists.cpp [hunk 9]
```diff
@@ -19,6 +19,10 @@
 #include <faiss/index_io.h>
 #include <faiss/IVFlib.h>
 
+#ifdef _MSC_VER
+#define drand48 rand
```
Comment (mdouze) [inline faiss/tests/test_dealloc_invlists.cpp:23]: drand48 returns a random number in [0, 1], rand is an integer

File: faiss/tests/test_sliding_ivf.cpp [hunk 10]
```diff
@@ -19,6 +19,11 @@
 #include <faiss/clone_index.h>
 #include <faiss/IVFlib.h>
 
+#ifdef _MSC_VER
+#define drand48 rand
```
Comment (mdouze) [inline faiss/tests/test_sliding_ivf.cpp:23]: plase replace with the std C++ random generator

File: faiss/utils/utils.cpp [hunk 11]
```diff
@@ -173,25 +191,6 @@ void reflection_ref (const float * u, float * x, size_t n, size_t d, size_t nu)
  * Some matrix manipulation functions
  ***************************************************************************/
 
-
```
Comment (mdouze) [inline faiss/utils/utils.cpp:176]: why did you move this away?

Comment (beauby) [reply faiss/utils/utils.cpp:176]: Declaration was in `distances.h` (not included here, which led to link errors on Windows).

**Comment by beauby** [comment]
Closing this as all commits have been exportedd in separate PRs.
